src/Adapter/Product/CommandHandler/UpdateProductSuppliersHandler.php

```diff
@@ -0,0 +1,325 @@
+                ),
+                CannotUpdateProductException::FAILED_UPDATE_DEFAULT_SUPPLIER
+            );
+        }
+
+        if ($defaultSupplierIsProvided) {
+            $this->updateDefaultSupplier($productId, $command->getDefaultSupplierId()->getValue());
+        }
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param ProductSupplier[] $productSuppliers
+     *
+     * @throws CannotDeleteProductSupplierException
+     * @throws CombinationConstraintException
+     * @throws CurrencyException
+     * @throws ProductSupplierException
+     * @throws SupplierException
+     */
+    private function updateProductSuppliers(ProductId $productId, array $productSuppliers): void
+    {
+        $deletableProductSupplierIds = $this->getDeletableProductSupplierIds($productId, $productSuppliers);
+
+        foreach ($productSuppliers as $productSupplier) {
+            if ($productSupplier->getProductSupplierId()) {
+                $this->updateProductSupplier($productSupplier);
+            } else {
+                $this->addProductSupplier($productId, $productSupplier);
+            }
+        }
+
+        $this->deleteProductSuppliers($deletableProductSupplierIds);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param ProductSupplier $productSupplier
+     *
+     * @return ProductSupplierId
+     *
+     * @throws CurrencyException
+     * @throws CombinationConstraintException
+     * @throws SupplierException
+     */
+    private function addProductSupplier(ProductId $productId, ProductSupplier $productSupplier): ProductSupplierId
+    {
+        $combinationId = $productSupplier->getCombinationId();
+
+        if ($combinationId === CombinationId::NO_COMBINATION) {
+            $combinationId = null;
+        }
+
+        $command = new AddProductSupplierCommand(
+            $productId->getValue(),
+            $productSupplier->getSupplierId(),
+            $productSupplier->getCurrencyId(),
+            $productSupplier->getReference(),
+            $productSupplier->getPriceTaxExcluded(),
+            $combinationId
+        );
+
+        return $this->addProductSupplierHandler->handle($command);
+    }
+
+    /**
+     * @param ProductSupplier $productSupplier
+     */
+    private function updateProductSupplier(ProductSupplier $productSupplier): void
+    {
+        $command = new UpdateProductSupplierCommand($productSupplier->getProductSupplierId());
+        $command->setCurrencyId($productSupplier->getCurrencyId())
+            ->setReference($productSupplier->getReference())
+            ->setPriceTaxExcluded($productSupplier->getPriceTaxExcluded())
+        ;
+
+        if ($productSupplier->getCombinationId()) {
+            $command->setCombinationId($productSupplier->getCombinationId());
+        }
+
+        $this->updateProductSupplierHandler->handle($command);
+    }
+
+    /**
+     * @param ProductId $productId
+     * @param int $defaultSupplierId
+     *
```

Comment:
Maybe a little comment in this phpDoc could explain the complex data structure for default supplier (with combinations, without ...) ? to help clear the confusion on this topic :)